File: src/voice/tts_integration.py
"""
Voice AI Assistant - ElevenLabs TTS Integration
Provides text-to-speech conversion with streaming support
"""

import asyncio
import tempfile
import time
import os
import subprocess
from pathlib import Path
from typing import AsyncGenerator, Tuple

from elevenlabs import ElevenLabs

from src.config import config


def get_elevenlabs_client() -> ElevenLabs:
    """Get a configured ElevenLabs client"""
    return ElevenLabs(api_key=config.elevenlabs.api_key)


async def text_to_speech(
    text: str,
    voice_id: str = None,
    model_id: str = None
) -> Tuple[bytes, float]:
    """
    Convert text to speech and return audio bytes
    
    Returns:
        Tuple of (audio_bytes, latency_ms)
    """
    start_time = time.time()
    
    client = get_elevenlabs_client()
    voice_id = voice_id or config.elevenlabs.voice_id
    model_id = model_id or config.elevenlabs.model
    
    # Run in executor to avoid blocking
    loop = asyncio.get_event_loop()
    
    audio_generator = await loop.run_in_executor(
        None,
        lambda: client.text_to_speech.convert(
            voice_id=voice_id,
            text=text,
            model_id=model_id,
        )
    )
    
    # Collect audio chunks
    chunks = []
    for chunk in audio_generator:
        chunks.append(chunk)
    
    audio_bytes = b''.join(chunks)
    latency_ms = (time.time() - start_time) * 1000
    
    return audio_bytes, latency_ms


async def stream_text_to_speech(
    text: str,
    voice_id: str = None,
    model_id: str = None
) -> AsyncGenerator[bytes, None]:
    """
    Stream text to speech for lower latency
    
    Yields:
        Audio chunks as they become available
    """
    client = get_elevenlabs_client()
    voice_id = voice_id or config.elevenlabs.voice_id
    model_id = model_id or config.elevenlabs.model
    
    loop = asyncio.get_event_loop()
    
    audio_generator = await loop.run_in_executor(
        None,
        lambda: client.text_to_speech.convert(
            voice_id=voice_id,
            text=text,
            model_id=model_id,
        )
    )
    
    for chunk in audio_generator:
        yield chunk


async def save_audio_to_file(audio_bytes: bytes, filepath: str | Path) -> None:
    """Save audio bytes to a file"""
    filepath = Path(filepath)
    
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(
        None,
        lambda: filepath.write_bytes(audio_bytes)
    )


async def play_audio(audio_bytes: bytes) -> None:
    """Play audio bytes (saves to temp file and plays with system player)"""
    # Save to temp file
    with tempfile.NamedTemporaryFile(suffix='.mp3', delete=False) as f:
        f.write(audio_bytes)
        temp_path = f.name
    
    print(f"🔊 Audio saved to: {temp_path}")
    
    # Try to play using system command
    try:
        if os.name == 'nt':  # Windows
            subprocess.run(['start', '', temp_path], shell=True, check=False)
        elif os.uname().sysname == 'Darwin':  # macOS
            subprocess.run(['afplay', temp_path], check=False)
        else:  # Linux
            subprocess.run(['aplay', temp_path], check=False)
    except Exception as e:
        print(f"Could not auto-play audio: {e}")
        print(f"Audio file saved at: {temp_path}")


async def get_voices() -> list[dict]:
    """Get available voices from ElevenLabs"""
    client = get_elevenlabs_client()
    
    loop = asyncio.get_event_loop()
    response = await loop.run_in_executor(
        None,
        lambda: client.voices.get_all()
    )
    
    return [
        {
            'voice_id': v.voice_id,
            'name': v.name or 'Unknown',
        }
        for v in response.voices
    ]

File: src/voice/ttsIntegration.ts
import { ElevenLabsClient } from 'elevenlabs';
import { config } from '../config/index.js';

let elevenlabsClient: ElevenLabsClient | null = null;

/**
 * Get or create the ElevenLabs client singleton
 */
function getElevenLabsClient(): ElevenLabsClient {
    if (!elevenlabsClient) {
        elevenlabsClient = new ElevenLabsClient({
            apiKey: config.elevenlabs.apiKey,
        });
    }
    return elevenlabsClient;
}

/**
 * Convert text to speech and return audio buffer
 */
export async function textToSpeech(
    text: string,
    options: {
        voiceId?: string;
        modelId?: string;
    } = {}
): Promise<{ audio: Buffer; latencyMs: number }> {
    const startTime = Date.now();

    const client = getElevenLabsClient();

    const audioResponse = await client.textToSpeech.convert(
        options.voiceId || config.elevenlabs.voiceId,
        {
            text,
            model_id: options.modelId || config.elevenlabs.model,
        }
    );

    // Collect chunks into a buffer
    const chunks: Uint8Array[] = [];

    // Handle the response based on its type
    if (audioResponse && typeof audioResponse[Symbol.asyncIterator] === 'function') {
        for await (const chunk of audioResponse as AsyncIterable<Uint8Array>) {
            chunks.push(chunk);
        }
    } else if (audioResponse instanceof ArrayBuffer) {
        chunks.push(new Uint8Array(audioResponse));
    }

    // Combine chunks
    const totalLength = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
    const combined = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of chunks) {
        combined.set(chunk, offset);
        offset += chunk.length;
    }

    return {
        audio: Buffer.from(combined),
        latencyMs: Date.now() - startTime,
    };
}

/**
 * Stream text to speech for lower latency
 * Returns chunks of audio as they become available
 */
export async function* streamTextToSpeech(
    text: string,
    options: {
        voiceId?: string;
        modelId?: string;
    } = {}
): AsyncGenerator<Buffer> {
    const client = getElevenLabsClient();

    const audioResponse = await client.textToSpeech.convert(
        options.voiceId || config.elevenlabs.voiceId,
        {
            text,
            model_id: options.modelId || config.elevenlabs.model,
        }
    );

    if (audioResponse && typeof audioResponse[Symbol.asyncIterator] === 'function') {
        for await (const chunk of audioResponse as AsyncIterable<Uint8Array>) {
            yield Buffer.from(chunk);
        }
    } else if (audioResponse instanceof ArrayBuffer) {
        yield Buffer.from(audioResponse);
    }
}

/**
 * Get available voices
 */
export async function getVoices(): Promise<Array<{ voice_id: string; name: string }>> {
    const client = getElevenLabsClient();
    const response = await client.voices.getAll();

    return response.voices.map((v: { voice_id: string; name?: string }) => ({
        voice_id: v.voice_id,
        name: v.name || 'Unknown',
    }));
}

/**
 * Utility to save audio to file
 */
export async function saveAudioToFile(
    audio: Buffer,
    filepath: string
): Promise<void> {
    const fs = await import('fs/promises');
    await fs.writeFile(filepath, audio);
}

/**
 * Play audio (requires a player - mainly for testing)
 */
export async function playAudio(audio: Buffer): Promise<void> {
    const path = await import('path');
    const os = await import('os');
    const tempPath = path.join(os.tmpdir(), `voice-ai-${Date.now()}.mp3`);

    await saveAudioToFile(audio, tempPath);
    console.log(`🔊 Audio saved to: ${tempPath}`);

    // Try to play using system command
    const { exec } = await import('child_process');
    const { promisify } = await import('util');
    const execAsync = promisify(exec);

    try {
        if (process.platform === 'win32') {
            await execAsync(`start "" "${tempPath}"`);
        } else if (process.platform === 'darwin') {
            await execAsync(`afplay "${tempPath}"`);
        } else {
            await execAsync(`aplay "${tempPath}" || mpv "${tempPath}" || play "${tempPath}"`);
        }
    } catch {
        console.log('Could not auto-play audio. File saved at:', tempPath);
    }
}
